### ISSUE TITLE
Increased / more verbose type aliases for improved readability of user defined content

### ISSUE THREAD

**Issue by dsm-72** [pytorch/pytorch#103109]
### 🚀 The feature, motivation and pitch

Right now PyTorch exports a few classes and has a few type aliases under `torch.types`. 
While these are useful, for those working with pytorch there may be a case to use `TypeAlias` to help clarify arguments and return statements. For example consider just a few of the following aliases:

```python
TorchDevice: TypeAlias = Union[Device, TorchDeviceTypes]


TorchTensor: TypeAlias = Union[tuple(torch._tensor_classes)] # could be just torch.Tensor, but doesn't offer all the hints
TorchTensors: TypeAlias = Sequence[TorchTensor]
TorchLayer: TypeAlias = torch.nn.Module
TorchLayers: TypeAlias = Sequence[TorchLayer]

TorchLoss: TypeAlias = torch.nn.Module
TorchLosses: TypeAlias = Sequence[TorchLoss]

StateDict: TypeAlias = dict
MaskTensor: TypeAlias = Union[torch.BoolTensor, torch.IntTensor]

HiddenState: TypeAlias = TorchTensor
CellState: TypeAlias = TorchTensor

GRUState: TypeAlias = HiddenState
LSTMStates: TypeAlias = Tuple[HiddenState, CellState]

RNNStates: TypeAlias = Union[GRUState, LSTMStates]
```

Suppose someone is defining custom layer that users can initialize with either a `GRU` or `LSTM`.  Showing that the forward call returns `Tuple[Tensor, RNNStates]` may help developer more readily remember that `RNNStates` is either a single `tensor`, or a `tuple`, which could otherwise be overlooked. 

Likewise `TorchLayer` and `TorchLoss`, while both an alias for `nn.Module` would help clarify if someone is passing a loss function or a layer into a model a wider variety of what expect input is. 

Does PyTorch need an `torch.aliases` submodule  with a bunch of different aliases to use? Not technically. I do, however, find working with others on projects where doing so can make things much more clear for collaboration. One could also just use longer type annotations, but I find those eventually clutter code e.g.

```
def foo(..., rnn_states: Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]], ...):
    pass
```

It is just a preference.

### Additional context

I put it in a [PR](https://github.com/pytorch/pytorch/pull/102914#issuecomment-1578935718)

**Comment by jbschlosser** [comment]
Added triage review for discussion.

**Comment by dsm-72** [comment]
@malfet what do you think?

**Comment by dsm-72** [comment]
Relevant would be [`np.typing`][nptyping]

[nptyping]: https://pypi.org/project/nptyping/

**Comment by ezyang** [comment]
Having some more aliases is generally good. Bikeshedding out what exactly the aliases should be is hard though. Some of the aliases proposed here are redundant with aliases we already have; for example, TorchDevice is just torch.types.Device, as best I can tell. Annotations should not be prefixed with Torch as they are already in torch namespace, and it's not idiomatic with how we tend to define schema. Some of the proposed aliases don't seem very good to me (e.g., TorchLoss.)

**Comment by dsm-72** [comment]
@ezyang Thank you for your feedback. Yeah, bikeshedding is probably the appropriate word here :P This isn't high priority in any sense.

I make no claim as to know what they should be or that those suggestions above are good. I only am noting that they have been useful in a few projects.

Motivation regarding `TorchDevice` instead of `torch.types.Device` as well as the `Torch` prefix :
```python
# Personal preference that writing `torch.types` over and over again is not something I want to do
def do_something_with_device(device:torch.types.Device) -> torch.types.Device:

# Device is so generic it might clash with other types or aliases
from torch.types import Device

# Fine if you only need device, more irksome if you are personally prefixing Torch in front of 7-8 aliases that you might be using
from torch.types import Device as TorchDevice

# too generic
import torch.types as types

# maybe better?
import torch.types as ttypes

# maybe better still?
from torch.aliases import (PTDevice, PTTensor, PTOptim, PTLoss...)
```


Note about `torch.types.Device` (see [How to get list of all valid devices](https://github.com/pytorch/pytorch/issues/97026))
It may be useful to have aliases for Device literals:
```python
# either
DeviceName = Literal['cpu', 'cuda', 'ipu', 'xpu', 'mkldnn', 'opengl', 'opencl', 'ideep', 'hip', 've', 'fpga', 'ort', 'xla', 'lazy', 'vulkan', 'mps', 'meta', 'hpu', 'privateuseone' ]

# or
DeviceCPU = Literal['cpu']
DeviceMPS = Literal['cpu']
# ...
DeviceName = Union[DeviceCPU, DeviceMPS, ...]
```

As it can be useful when some methods (within torch or user defined) take a device argument.  


Motivation regarding `TorchLoss`. Per pytorch documentations it is not recommended that users subclass `_Loss` directly. So both custom layers and custom loss functions end up subclassing `nn.Module`. Here the alias just makes it clear that while the method is expecting a subclass of `nn.Module` that it will be used internally with the assumptions to go with `Loss` instances. 




Of interest is [nptyping](https://pypi.org/project/nptyping/) which takes a different approach

```python
>>> from nptyping import NDArray, Int, Shape

>>> arr: NDArray[Shape["2, 2"], Int]
```

so maybe that is the better solution, something like:

```python
import torch, torch.nn as nn
from torch.typing import PTTensor, Dims, Float, Device

class MyMod(nn.Module):
    def forward(x:PTTensor[Dims[3], Float, Device['mps']]):
        pass
```

**Comment by ezyang** [comment]
> Motivation regarding TorchDevice instead of torch.types.Device as well as the Torch prefix :

Yeah, my claim is if you're in a PyTorch project there actually really is only one Device of interest. And if you're in some weird situation where there are multiple devices, you can just realias it yourself.

> It may be useful to have aliases for Device literals:

Seems pretty rare to have functions that only take one device type

> Motivation regarding TorchLoss. Per pytorch documentations it is not recommended that users subclass _Loss directly. So both custom layers and custom loss functions end up subclassing nn.Module. Here the alias just makes it clear that while the method is expecting a subclass of nn.Module that it will be used internally with the assumptions to go with Loss instances.

Meh, I guess I'd be more keen on it if we actually had shapes (because then you could mark the output of Loss as scalar). Seems like marginal benefit, esp because there is no way to actually check in mypy that a type is Loss.

> Of interest is [nptyping](https://pypi.org/project/nptyping/) which takes a different approach

nptyping is definitely much more in the style of what I'd be hoping for!

**Comment by dsm-72** [comment]
@ezyang it is fairly easy to implement and a quick pypi check (https://pypi.org/project/torchtyping/)


```python
import torch
from typing import TypeVar, Generic, Union
from torch.optim import Optimizer
from torch.nn.modules.loss import _Loss
from torch.nn import Module


DType = TypeVar('DType', 'Float', 'Double', 'Half', 'Byte', 'Char', 'Short', 'Int', 'Long', 'Bool')
Device = TypeVar('Device', 'cpu', 'cuda', 'ipu', 'xpu', 'mkldnn', 'opengl', 'opencl', 'ideep', 'hip', 've', 'fpga', 'ort', 'xla', 'lazy', 'vulkan', 'mps', 'meta', 'hpu', 'privateuseone')

class Dims(Generic[DType, Device]):
    def __init__(self, *args: Union[int, str]):
        self.shape = args


class PTTensor(Generic[Dims, DType, Device]):
    def __init__(self, dims: Dims, dtype: DType, device: Device):
        self.dims = dims
        self.dtype = dtype
        self.device = device

class NamedDims(Generic[DType, Device]):
    def __init__(self, **kwargs: Union[int, str]):
        self.shape = kwargs

class PTTensor(Generic[NamedDims, DType, Device]):
    def __init__(self, dims: NamedDims, dtype: DType, device: Device):
        self.dims = dims
        self.dtype = dtype
        self.device = device

class Batched(Generic[PTTensor]):
    def __init__(self, tensor: PTTensor):
        self.tensor = PTTensor(NamedDims(*tensor.dims.shape, batch=1), tensor.dtype, tensor.device)

class BatchedQ(Generic[PTTensor]):
    def __init__(self, tensor: PTTensor):
        self.tensor = Union[tensor, Batched[tensor]]

class PTOptim(Generic[Optimizer]):
    def __init__(self, optim: Optimizer):
        self.optim = optim

class PTLoss(Generic[_Loss]):
    def __init__(self, loss: _Loss):
        self.loss = loss

class PTLayer(Generic[Module]):
    def __init__(self, layer: Module):
        self.layer = layer

```

so something like:

```python
import torch, torch.nn as nn
from pttyping import PTTensor, NamedDims, DType, Device, Batched, BatchedQ, PTOptim, PTLoss, PTLayer

class MyMod(nn.Module):
    def forward(self, x: BatchedQ[PTTensor[NamedDims(H=3, W=224, C=224), DType['Float'], Device['cpu']]]):
        pass

    def backward(self, loss: PTLoss[nn.CrossEntropyLoss], optim: PTOptim[torch.optim.SGD]):
        pass

class MyConv(PTLayer[nn.Conv2d]):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1)

```
